packages/alfa-table/src/table.ts

```diff
@@ -1,684 +1,854 @@
+     * Carry out step 9.1 of the table forming algorithm.
+     *
+     * @see https://html.spec.whatwg.org/#forming-a-table
+     */
+    function processColumnGroup(element: Element): void {
+      let children = element
+        .children()
+        .filter(isElement)
+        .filter(hasName("col"));
+
+      if (!children.isEmpty()) {
+        // 1
+        const xStart = xWidth;
+
+        // 2
+        // Nothing to do
+
+        // Steps 3-6 are repeated for as long as there are children left in the
+        // column group.
+        while (current().isSome()) {
+          // 3
+          const span = integerValue(
+            current().get(),
+            "span",
+            1 /* lower */,
+            1000 /* upper */
+          );
+
+          // 4
+          xWidth += span;
+
+          // 5
+          // Nothing to do
+
+          // 6
+          advance();
         }
+
+        // 7
+        const group = Column.group(xStart, xWidth - xStart);
+
+        groups.push(group);
+
+        // Todo
+      } else {
+        // 1
+        const span = integerValue(
+          current().get(),
+          "span",
+          1 /* lower */,
+          1000 /* upper */
+        );
+
+        // 2
+        xWidth += span;
+
+        // 3
+        const group = Column.group(xWidth - span, span);
+
+        groups.push(group);
+
+        // Todo
       }
 
-      return this.updateCells({
-        cells: List.from(
-          map(this.cells, (cell) => {
-            // We need to know if the cell is covering a slot with a data cell in the same column/row
-            let dataInColumns = false;
-            let dataInRows = false;
-            for (let x = cell.anchor.x; x < cell.anchor.x + cell.width; x++) {
-              if (columnHasData[x] ?? false) {
-                dataInColumns = true;
-              }
-            }
-            for (let y = cell.anchor.y; y < cell.anchor.y + cell.height; y++) {
-              if (rowHasData[y] ?? false) {
-                dataInRows = true;
-              }
-            }
+      /**
+       * Get the current child element of the column group.
+       */
+      function current(): Option<Element> {
+        return children.first();
+      }
 
-            return cell.addHeaderVariant(dataInColumns, dataInRows);
-          })
-        ),
-      });
+      /**
+       * Advance to the next child element of the column group.
+       */
+      function advance(): Option<Element> {
+        children = children.rest();
+        return current();
+      }
     }
 
-    public findGroupHeaders(): Builder {
-      const rowGroupHeaders: Array<Cell.Builder> = [];
-      const columnGroupHeaders: Array<Cell.Builder> = [];
+    /**
+     * For all downwards growing cells, extend their height to the current
+     * y-coordinate.
+     *
+     * @see https://html.spec.whatwg.org/#algorithm-for-growing-downward-growing-cells
+     */
+    function growCells(): void {
+      for (let [cell, i] of downwardGrowing) {
```

Comment:
I'm wondering if `Cell` should have an `id` field that holds `i` since it is essentially what this is.
This would ensure the coupling between the cell and its index and could limit problems in building the `[cell, i]` tuple.

Cells definitely do not have an "index in an array", but it does make sense for them to have an ID (to track a cell through its various instances), and it is just serendipity that Tables are using that ID as an index in an array (or key in an object)…

Reply:
These indices have no use outside of `fromTable()` and so shouldn't leak into the `Cell` class. Beyond that, they're uniquely identifiable within a table by their anchor.